scc/light_client/retry.go

```diff
@@ -0,0 +1,115 @@
+package light_client
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/0xsoniclabs/sonic/scc"
+	"github.com/0xsoniclabs/sonic/scc/cert"
+	"github.com/Fantom-foundation/lachesis-base/inter/idx"
+)
+
+// retryP is a provider that retries requests a certain number of times on failed
+// requests for certificates with a determined wait between retries. It is used
+// to wrap other providers to add a retry mechanism.
+//
+// Fields:
+// - provider: The underlying provider to retry requests on.
+// - attempts: The maximum number of attempt to ask for certificates.
+// - delay: The time to wait between retries.
+type retryP struct {
+	provider provider
+	attempts uint
+	delay    time.Duration
+}
+
+// newRetry creates a new retryP provider with the given provider and maximum
+// number of retries and delay between retries.
+//
+// Parameters:
+// - provider: The underlying provider to wrap with retry logic.
+// - attempts: The maximum number of attempt to ask for certificates.
+// - delay: The time to wait between retries.
+//
+// Returns:
+// - *retryP: A new retryP provider instance.
+func newRetry(provider provider, attempts uint, delay time.Duration) *retryP {
+	return &retryP{
+		provider: provider,
+		attempts: attempts,
+		delay:    delay,
+	}
+}
+
+// Close closes the retryP provider.
+// Closing an already closed provider has no effect.
+func (r *retryP) close() {
+	r.provider.close()
+}
+
+// getCommitteeCertificates returns up to `maxResults` consecutive committee
+// certificates starting from the given period.
+//
+// Parameters:
+// - first: The starting period for which to retrieve committee certificates.
+// - maxResults: The maximum number of committee certificates to retrieve.
+//
+// Returns:
+// - []cert.CommitteeCertificate: A slice of committee certificates.
+// - error: An error if the provider failed to obtain the requested certificates.
+func (r retryP) getCommitteeCertificates(first scc.Period, maxResults uint64) ([]cert.CommitteeCertificate, error) {
+	result, err := retry(r.attempts, r.delay, func() (any, error) {
+		return r.provider.getCommitteeCertificates(first, maxResults)
+	})
+	if err != nil {
+		return nil, err
+	}
+	return result.([]cert.CommitteeCertificate), nil
```

Comment:
The `retry` function is generic, so why mixing generics with reflection?
```suggestion
	return retry(r.attempts, r.delay, func() ([]cert.CommitteeCertificate, error) {
		return r.provider.getCommitteeCertificates(first, maxResults)
	})
```

Reply:
https://github.com/0xsoniclabs/sonic/pull/128/commits/91dfbe1c5d1a763e2d73c147f8e929977c9b9f00